cmd/kubeadm/app/cmd/token.go

```diff
@@ -277,22 +296,85 @@ func RunListTokens(out io.Writer, errW io.Writer, client clientset.Interface) er
 		return errors.Wrap(err, "failed to list bootstrap tokens")
 	}
 
-	w := tabwriter.NewWriter(out, 10, 4, 3, ' ', 0)
-	fmt.Fprintln(w, "TOKEN\tTTL\tEXPIRES\tUSAGES\tDESCRIPTION\tEXTRA GROUPS")
-	for _, secret := range secrets.Items {
+	if outputFormat == "text" {
+		w := tabwriter.NewWriter(out, 10, 4, 3, ' ', 0)
+		fmt.Fprintln(w, "TOKEN\tTTL\tEXPIRES\tUSAGES\tDESCRIPTION\tEXTRA GROUPS")
+		for _, secret := range secrets.Items {
 
-		// Get the BootstrapToken struct representation from the Secret object
-		token, err := kubeadmapi.BootstrapTokenFromSecret(&secret)
-		if err != nil {
-			fmt.Fprintf(errW, "%v", err)
-			continue
+			// Get the BootstrapToken struct representation from the Secret object
+			token, err := kubeadmapi.BootstrapTokenFromSecret(&secret)
+			if err != nil {
+				fmt.Fprintf(errW, "%v", err)
+				continue
+			}
+
+			// Get the human-friendly string representation for the token
+			humanFriendlyTokenOutput := humanReadableBootstrapToken(token)
+			fmt.Fprintln(w, humanFriendlyTokenOutput)
 		}
+		w.Flush()
+	} else if outputFormat == "yaml" || outputFormat == "json" {
+		var items []*outputItem
+		for _, secret := range secrets.Items {
+			// Get the BootstrapToken struct representation from the Secret object
+			token, err := kubeadmapi.BootstrapTokenFromSecret(&secret)
+			if err != nil {
+				fmt.Fprintf(errW, "%v", err)
+				continue
+			}
+			clusterConfig, err := cmdutil.GetClusterConfig(kubeConfigFile)
+			if err != nil {
+				return errors.Wrapf(err, "failed to load cluser config from %s", kubeConfigFile)
+			}
+
+			publicKeyPins, err := cmdutil.GetCAPubKeyPins(clusterConfig)
+			if err != nil {
+				return errors.Wrapf(err, "failed to get CA certs hashes from cluster config")
+			}
 
-		// Get the human-friendly string representation for the token
-		humanFriendlyTokenOutput := humanReadableBootstrapToken(token)
-		fmt.Fprintln(w, humanFriendlyTokenOutput)
+			ttl := "<forever>"
+			expires := "<never>"
+			if token.Expires != nil {
+				ttl = duration.ShortHumanDuration(token.Expires.Sub(time.Now()))
+				expires = token.Expires.Format(time.RFC3339)
+			}
+
+			items = append(
+				items,
+				&outputItem{
+					Description:          token.Description,
+					Expires:              expires,
+					TTL:                  ttl,
+					Token:                token.Token,
+					Usages:               token.Usages,
+					Groups:               token.Groups,
+					ControlPlaneHostPort: strings.Replace(clusterConfig.Server, "https://", "", -1),
+					PublicKeyPins:        publicKeyPins,
+				},
+			)
+		}
+		var output string
+		jsn, err := json.MarshalIndent(items, "", "  ")
```

Comment:
can we move this in the branch bellow?
[1]

Reply:
we can't as its result is used in both branches.

Reply:
It was actually a failed attempt to get the order identical for json and yaml.
Now the order for JSON is the same as defined in the structure:
```
{
    "Token": "0djuow.qjyjiju21mhxke3r",
    "Description": "",
    "Expires": "2019-03-25T19:17:43+02:00",
    "TTL": "23h",
    "Usages": [
      "authentication",
      "signing"
    ],
    "Groups": [
      "system:bootstrappers:kubeadm:default-node-token"
    ],
    "PublicKeyPins": [
      "sha256:4216f79009fb8043f366d63eccb71249b5a09de36e8d9b249fae19b5fb6a7591"
    ],
    "ControlPlaneHostPort": "10.237.72.179:6443"
  },
```

However, for yaml it's changed for some reason (sorted alphabetically):
```
- ControlPlaneHostPort: 10.237.72.179:6443
  Description: ""
  Expires: "2019-03-25T19:17:43+02:00"
  Groups:
  - system:bootstrappers:kubeadm:default-node-token
  PublicKeyPins:
  - sha256:4216f79009fb8043f366d63eccb71249b5a09de36e8d9b249fae19b5fb6a7591
  TTL: 23h
  Token: 0djuow.qjyjiju21mhxke3r
  Usages:
  - authentication
  - signing
```

I'm still wondering if it's possible to have the Yaml order the same as JSON.

Reply:
in that case please replace these comments:
```
//jsn, err := json.MarshalIndent(items, "", "  ")
			//if err != nil {
			//	return err
			//}
```

with a single line comment explaining that we just use `string(jsn)`

Reply:
Done